File: app/src/main/java/com/myapplication/rick/packagedelivery/RouteFormat.java
package com.myapplication.rick.packagedelivery;

import android.support.annotation.NonNull;

import java.util.ArrayList;

/**
 * Created by Rick on 1-5-2016.
 */
class RouteFormat {
    private final ArrayList<Street> route;
    public RouteFormat() {
        route = new ArrayList<>(75);
    }

    public RouteFormat(@NonNull ArrayList<Street> route) {
        this.route = route;
    }

    public ArrayList<Street> getRoute() {
        return route;
    }

    //if street is not null and route does not contain the street, add
    public boolean addStreet(Street street) {
        return street != null && !route.contains(street) && route.add(street);
    }

    /*
    public boolean removeStreet(Street street) {
        return route.remove(street);
    }*/

    public int totalStreets() {
        return route.size();
    }

    public void reset() {
        route.clear();
    }

    public ArrayList<String[]> toCSVList() {
        ArrayList<String[]> csvStreets = new ArrayList<>();
        for (Street street : route) {
            csvStreets.add(street.toCSV());
        }
        return csvStreets;
    }
}

File: app/src/main/java/com/myapplication/rick/packagedelivery/RouteFormat.kt
package com.myapplication.rick.packagedelivery

import android.os.Parcel
import android.os.Parcelable
import java.util.ArrayList

/**
 * Created by Rick on 1-5-2016.
 * Sorted list describing the order of a route
 */
internal class RouteFormat(val name: String, val route: ArrayList<Street> = ArrayList(75)): Parcelable {
    init {
        if (name.isEmpty()) throw IllegalArgumentException("Name mustn't be empty")
    }

    constructor(parcel: Parcel) : this(
            parcel.readString(),
            arrayListOf<Street>().apply {
                parcel.readList(this, Street::class.java.classLoader)
            })

    fun addStreet(street: Street): Boolean =
            !route.contains(street) && route.add(street)

    fun removeStreet(street: Street): Boolean = route.remove(street)

    fun totalStreets(): Int = route.size

    fun reset() = route.clear()

    fun toCSVList(): ArrayList<Array<String>> = route.mapTo(ArrayList()) { it.toCSV() }

    override fun toString(): String = name
    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(name)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<RouteFormat> {
        override fun createFromParcel(parcel: Parcel): RouteFormat = RouteFormat(parcel)

        override fun newArray(size: Int): Array<RouteFormat?> = arrayOfNulls(size = size)
    }
}
